Bleh. Looks like there is strangeness now when tracing a non-PIE exe. Specifically, it doesn't map correctly? Here's a logging example:

```python In [1]: import logging

In [2]: logging.basicConfig(level=logging.INFO)

In [3]: import tracer
WARNING:claripy:Claripy is setting the recursion limit to 15000. If Python segfaults, I am sorry.
INFO:simuvex.plugins.unicorn:native plugin is enabled

In [4]: t = tracer.Tracer("./login","test\ntest2\n")
INFO:angr.project:Loading binary ./login
INFO:cle.loader:[Rebasing /home/user/fgets_bug/login @0x0]
INFO:cle.loader:[Rebasing /lib/i386-linux-gnu/libc-2.24.so @0x9000000]
INFO:cle.loader:[Rebasing /lib/i386-linux-gnu/ld-2.24.so @0xa000000]
INFO:cle.loader:[Rebasing ##cle_tls## @0xb000000]
INFO:cle.loader:[Rebasing ##angr_externs## @0xc000000]
INFO:cle.loader:[Rebasing ##angr_externs## @0xd000000]
INFO:angr.project:Providing strcmp from libc.so.6 with SimProcedure
INFO:angr.project:Providing printf from libc.so.6 with SimProcedure
INFO:angr.project:Providing fgets from libc.so.6 with SimProcedure
INFO:angr.project:Providing __stack_chk_fail from libc.so.6 with SimProcedure
INFO:angr.project:Providing puts from libc.so.6 with SimProcedure
INFO:angr.project:Providing __libc_start_main from libc.so.6 with SimProcedure
INFO:angr.path_history:Callstack on the path is unbalanced.
INFO:angr.simos:A factory.call_state-created path returned!
INFO:angr.path_history:Callstack on the path is unbalanced.
INFO:angr.simos:A factory.call_state-created path returned!
INFO:angr.path_history:Callstack on the path is unbalanced.
INFO:angr.simos:A factory.call_state-created path returned!
INFO:tracer.Tracer:tracing as raw input
INFO:tracer.Tracer:trace consists of 61 basic blocks
Requesting 0x8048000
INFO:angr.project:Loading binary ./login
INFO:cle.loader:[Rebasing /home/user/fgets_bug/login @0x8048000]
INFO:cle.loader:[Rebasing /lib/i386-linux-gnu/libc-2.24.so @0x11000000]
INFO:cle.loader:[Rebasing /lib/i386-linux-gnu/ld-2.24.so @0x12000000]
INFO:cle.loader:[Rebasing ##cle_tls## @0x13000000]
INFO:cle.loader:[Rebasing ##angr_externs## @0x14000000]
INFO:cle.loader:[Rebasing ##angr_externs## @0x15000000]
INFO:angr.project:Providing strcmp from libc.so.6 with SimProcedure
INFO:angr.project:Providing printf from libc.so.6 with SimProcedure
INFO:angr.project:Providing fgets from libc.so.6 with SimProcedure
INFO:angr.project:Providing __stack_chk_fail from libc.so.6 with SimProcedure
INFO:angr.project:Providing puts from libc.so.6 with SimProcedure
INFO:angr.project:Providing __libc_start_main from libc.so.6 with SimProcedure
INFO:angr.path_history:Callstack on the path is unbalanced.
INFO:angr.simos:A factory.call_state-created path returned!
INFO:angr.path_history:Callstack on the path is unbalanced.
INFO:angr.simos:A factory.call_state-created path returned!
INFO:angr.path_history:Callstack on the path is unbalanced.
INFO:angr.simos:A factory.call_state-created path returned!
Got 0x10090000

In [5]: t._p.loader.all_objects()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-1c72fc263365> in <module>()
----> 1 t._p.loader.all_objects()

TypeError: 'list' object is not callable

In [6]: t._p.loader.all_objects
Out[6]: 
[<ELF Object login, maps [0x10090000:0x10092047]>,
 <ELF Object libc-2.24.so, maps [0x11000000:0x111b9a1b]>,
 <ELF Object ld-2.24.so, maps [0x12000000:0x12024917]>,
 <ELFTLSObj Object ##cle_tls##, maps [0x13000000:0x13030000]>,
 <AngrExternObject Object ##angr_externs##, maps [0x14000000:0x14004000]>,
 <AngrExternObject Object ##angr_externs##, maps [0x15000000:0x15004000]>]
```

The logging inside the loader function claims to give the main_bin the requested base_addr, but then afterwards it appears that is not the case. Any ideas on this?

--- a/tracer/tracer.py
+++ b/tracer/tracer.py
@@ -855,8 +855,7 @@ class Tracer(object):
 
     def _set_linux_simprocedures(self, project):
         for symbol in self.simprocedures:
-            project.set_sim_procedure(
-                    project.loader.main_bin,
+            project.hook_symbol(
                     symbol,
                     self.simprocedures[symbol])
 

--- a/tracer/tracer.py
+++ b/tracer/tracer.py
@@ -258,6 +258,7 @@ class Tracer(object):
                     # if so we need to take special care
                     r_plt = self._p.loader.main_bin.reverse_plt
                     if current.addr not in self._resolved \
+                            and self.previous is not None \
                             and self.previous.addr in r_plt:
                         self.bb_cnt += 2
                         self._resolved.add(current.addr)
@@ -1054,7 +1055,7 @@ class Tracer(object):
 
         self.remove_options |= so.simplification
         self.add_options |= options
-        entry_state = project.factory.entry_state(
+        entry_state = project.factory.full_init_state(
                 fs=fs,
                 concrete_fs=True,
                 chroot=self.chroot,

--- a/tracer/tracer.py
+++ b/tracer/tracer.py
@@ -1030,7 +1030,11 @@ class Tracer(object):
         prepare the initial paths for Linux binaries
         '''
 
-        project = angr.Project(self.binary,load_options={'main_opts': {'custom_base_addr': self.qemu_base_addr }})
+        # Only requesting custom base if this is a PIE
+        if self._p.loader.main_bin.pic:
+            project = angr.Project(self.binary,load_options={'main_opts': {'custom_base_addr': self.qemu_base_addr }})
+        else:
+            project = angr.Project(self.binary)
 
         if not self.crash_mode:
             self._set_linux_simprocedures(project)

